nearcore/src/migrations.rs

```diff
@@ -77,6 +80,90 @@ pub fn load_migration_data(chain_id: &str) -> MigrationData {
+        Self { config }
+    }
+}
+
+/// Asserts that node’s configuration does not use deprecated snapshot config
+/// options.
+///
+/// The `use_db_migration_snapshot` and `db_migration_snapshot_path`
+/// configuration options have been deprecated in favour of
+/// `store.migration_snapshot`.
+///
+/// This function panics if the old options are set and shows instruction how to
+/// migrate to the new options.
+///
+/// This is a hack which stops `StoreOpener` before it attempts migration.
+/// Ideally we would propagate errors nicely but that would complicate the API
+/// a wee bit so instead we’re panicking.  This shouldn’t be a big deal since
+/// the deprecated options are going away ‘soon’.
+fn assert_no_deprecated_config(config: &crate::config::NearConfig) {
+    use near_store::config::MigrationSnapshot;
+
+    let example = match (
+        config.config.use_db_migration_snapshot,
+        config.config.db_migration_snapshot_path.as_ref(),
+    ) {
+        (None, None) => return,
+        (Some(false), _) => MigrationSnapshot::Enabled(false),
+        (_, None) => MigrationSnapshot::Enabled(true),
+        (_, Some(path)) => MigrationSnapshot::Path(path.join("migration-snapshot")),
+    };
+    panic!(
+        "‘use_db_migration_snapshot’ and ‘db_migration_snapshot_path’ options \
+         are deprecated.\nSet ‘store.migration_snapshot’ to instead, e.g.:\n{}",
+        example.format_example()
+    )
+}
+
+impl<'a> near_store::StoreMigrator for Migrator<'a> {
+    type Err = anyhow::Error;
+    type Func = Self;
+
+    fn get_function(
+        &self,
+        version: DbVersion,
+    ) -> Result<&Self::Func, near_store::StoreMigratorError> {
+        assert_no_deprecated_config(self.config);
+        match version {
+            0..=26 => Err("1.26".into()),
+            DB_VERSION.. => unreachable!(),
+            _ => Ok(self),
+        }
+    }
+}
+
+impl<'a> near_store::StoreMigrationFunction for Migrator<'a> {
+    type Err = anyhow::Error;
+
+    fn migrate(&self, storage: &NodeStorage, version: DbVersion) -> Result<(), Self::Err> {
+        match version {
+            0..=26 => unreachable!(),
+            27 => {
+                // version 27 => 28: add DBCol::StateChangesForSplitStates
+                //
+                // Does not need to do anything since open db with option
+                // `create_missing_column_families`.  Nevertheless need to bump
+                // db version, because db_version 27 binary can't open
+                // db_version 28 db.  Combine it with migration from 28 to 29;
+                // don’t do anything here.
+                Ok(())
+            }
+            28 => near_store::migrations::migrate_28_to_29(storage).map_err(|err| err.into()),
+            29 => near_store::migrations::migrate_29_to_30(storage).map_err(|err| err.into()),
+            30 => migrate_30_to_31(storage, &self.config),
+            DB_VERSION.. => unreachable!(),
```

Comment:
Like this dispatch, very clean!

Reply:
One thing I don’t like is that this match on version needs to be repeated in get_function and also in migrate.  In first version I had it done differently where there was just one function but it involved returning a boxed closure which probably was too convoluted.

Reply:
I actually don't understand why we need a two-step process here. Is it so that  we can avoid creating snapshot if we won't be able to migrate anyway? Maybe just pass the minimum supported version as an integer together with the migrator? 

This doesn't solve the problem that we specify the minimum version twice, but seems to simplify things a bit. 


Or do we envision some other cases where get_function fails?

Reply:
> Is it so that we can avoid creating snapshot if we won't be able to migrate anyway?

Yes, that’s the only reason.  I think passing the minimum version as an integer is kinda worse because now that version is in completely different place in the code compared where the migrations are defined.  Another option is to have `is_supported` method which returns `Ok(())` or `Err(latest_release)`.  I kinda like that in current design you cannot call migrate without going through get_function, which wouldn’t be the case with `is_supported`, but perhaps that’s too much complication for too little gain?

Reply:
Yeah, my gut feeling is that that's too complicated. But I generally hate anything more complicated than C, should've been a Go programmer really :) I'd say as long as it is `dyn` and not impl, i am onboard. 


One option would be to pass `migrations: HashMap<DbVersion, Box<dyn Fn(&NodeStore)>>` perhaps?

Reply:
OK, I convinced myself that I’ve over-engineered this.  Going with a single trait with two methods.

Reply:
> One option would be to pass migrations: HashMap<DbVersion, Box<dyn Fn(&NodeStore)>> perhaps?

That was kind of my first version.  Specifically `get_migration_function(DbVersion) → Result<Box<dyn Fn(&NodeStorage)>, &str>` but I figured dealing with boxing closures is just a bit too convoluted.  I also prefer current approach with `match` to building a hash map because if DB_VERSION gets incremented, the `match` will stop compiling and it will have to migrator will have to be updated as well.  It wouldn’t quite work with a hash map.